spring-integration-core/src/main/java/org/springframework/integration/util/FunctionIterator.java

```diff
@@ -24,9 +26,10 @@
  * {@link #iterator} to a new object applying the {@link #function} on {@link #next()}.
  *
  * @author Artem Bilan
+ * @author Ruslan Stelmachenko
  * @since 4.1
  */
-public class FunctionIterator<T, V> implements Iterator<V> {
+public class FunctionIterator<T, V> implements Iterator<V>, Closeable {
```

Comment:
Why not `CloseableIterator` ?

Reply:
It is interesting. Actually first I do this, but then refactored, because `close()` method in `CloseableIterator` doesn't declare any exception, and `Closeable` does. If we implement `CloseableIterator` here, we should catch and ignore inner iterator's `IOException` or wrap it with runtime exception before throw inside `FunctionIterator.close()`.

Maybe I should just add `IOException` to `CloseableIterator.close`? What do you think?

The reason I removed this exception from `CloseableIterator.close` declaration is because the analogous interface from spring-data project also doesn't have it. I thought the intention was to make clear that iterator.close() should not throw `IOException` and simplify it's use, but I'm not sure.

What do you think?

Reply:
Good catch, indeed!

I think let's stick with the `Closeable` contract and have that `IOException` in our `close()` method as well! Although from here I think we won't need that method declaration in the `CloseableIterator` at all.